File: codes/julia/wk01_action_principle.jl
# Wk01 - Least action principle (Lagrangian mechanics)
# Physical Chemistry 2 - Prof. S. Joon Kwon - SPMDL - SKKU
using Plots, Printf

m, ω, T, qT = 1.0, 1.0, 2.0, 1.0
A = qT/sin(ω*T)
t = range(0, T, length=2001);  dt = step(t)
q_cl = A .* sin.(ω .* t)

grad(y) = [ (y[2]-y[1])/dt; (y[3:end].-y[1:end-2])./(2dt); (y[end]-y[end-1])/dt ]
trapz(y) = dt*(sum(y) - 0.5*(y[1]+y[end]))
action(q) = trapz(0.5m .* grad(q).^2 .- 0.5m*ω^2 .* q.^2)

S_cl = action(q_cl)
@printf("classical action S_cl = %.6f\n", S_cl)

p1 = plot(xlabel="t", ylabel="q(t)", title="Path family (fixed endpoints)", legend=false)
for e in range(-0.8, 0.8, length=9)
    plot!(p1, t, q_cl .+ e .* sin.(π .* t ./ T), c=:gray, lw=0.8)
end
plot!(p1, t, q_cl, c=:red, lw=2.2)
scatter!(p1, [0, T], [0, qT], c=:black)

p2 = plot(xlabel="ε", ylabel="S[q]", title="S minimized on classical path")
eps = range(-1, 1, length=81)
for n in 1:3
    S = [action(q_cl .+ e .* sin.(n*π .* t ./ T)) for e in eps]
    plot!(p2, eps, S, lw=1.5, label="mode n=$n")
end
scatter!(p2, [0.0], [S_cl], c=:black, label="classical")
display(plot(p1, p2, layout=(1,2), size=(1100,420)))
readline()

File: codes/cpp/wk01_action_principle.cpp
// Wk01 - Least action principle (Lagrangian mechanics)
// Physical Chemistry 2 - Prof. S. Joon Kwon - SPMDL - SKKU
// Build:  g++ -O2 -std=c++17 wk01_action_principle.cpp -o wk01_action_principle
// SHO: q(t) = q_cl(t) + eps*sin(n*pi*t/T)  ->  S(eps) minimized at eps = 0.
#include <cmath>
#include <cstdio>
#include <vector>

int main() {
    const double m = 1, w = 1, T = 2, qT = 1;
    const double A = qT/std::sin(w*T);
    const int Nt = 2000;
    const double dt = T/Nt;

    auto action = [&](double eps, int n) {
        double S = 0;
        for (int i = 0; i < Nt; ++i) {
            double t0 = i*dt, t1 = (i+1)*dt, tm = 0.5*(t0+t1);
            auto q = [&](double t){ return A*std::sin(w*t) + eps*std::sin(n*M_PI*t/T); };
            double qd = (q(t1) - q(t0))/dt;               // midpoint rule
            double qm = q(tm);
            S += (0.5*m*qd*qd - 0.5*m*w*w*qm*qm)*dt;
        }
        return S;
    };

    std::printf("classical action S_cl = %.6f\n\n", action(0.0, 1));
    std::printf("%8s  %12s  %12s  %12s\n", "eps", "S(n=1)", "S(n=2)", "S(n=3)");
    FILE* f = std::fopen("action_vs_eps.csv", "w");
    std::fprintf(f, "eps,S_n1,S_n2,S_n3\n");
    for (int k = 0; k <= 40; ++k) {
        double eps = -1.0 + 2.0*k/40;
        double s1 = action(eps,1), s2 = action(eps,2), s3 = action(eps,3);
        std::fprintf(f, "%.4f,%.8f,%.8f,%.8f\n", eps, s1, s2, s3);
        if (k % 5 == 0)
            std::printf("%8.3f  %12.6f  %12.6f  %12.6f\n", eps, s1, s2, s3);
    }
    std::fclose(f);
    std::printf("\nwrote action_vs_eps.csv — S(eps) is a parabola with min at eps=0\n");
    return 0;
}
